msu/hooks/entity/tactical/actor.nut

```diff
@@ -68,6 +68,27 @@
 		return ::Math.floor(getActionPointsMax());
 	}
 
+	// This function is called during `onExecute` of the `ai_pickup_weapon` behavior
+	// This hook is necessary to ensure that the correct items array is passed to the `payForAction` function
+	// TODO - needs discussion because in vanilla even if the AI picks up a weapon AND a shield, they only pay the cost ONCE i.e. 4 AP
+	// However with this change now they will be able to pick up a weapon with 0 Cost if they have Quick Hands - this conflicts with vanilla behavior
+	// We could leave this hook out completely but that will leave a loop hole in MSU implementation regarding skills which may affect AP cost of picking
+	// up items from the ground
+	local pickupMeleeWeaponAndShield = o.pickupMeleeWeaponAndShield;
```

Comment:
@TaroEld and @Enduriel Need your input on this "TODO" discussion point.

Reply:
How does this work for players? Afaik players can't do this so this is a bug in vanilla that we would be fixing.

Reply:
> How does this work for players? Afaik players can't do this so this is a bug in vanilla that we would be fixing.

This function is only for AI. And I think this isn't a bug but intended behavior in vanilla i.e. an AI entity is supposed to be able to pick up the gear (both weapon and shield) with a single item action.

Reply:
I don't like that ngl, I think we should maybe report this as a bug and see what overhype says?

Reply:
This function is only used for resurrecting skeletons and zombies in vanilla. Therefore I say that it is intended behavior (the resurrecting zombie/skeleton uses this function to pick up the gear on their tile). So, I suppose this function should be removed from our handling anyway?
```
Searching 4858 files for ".pickupMeleeWeaponAndShield("

D:\BB-Modding\vanilla\scripts\ai\tactical\behaviors\ai_pickup_weapon.nut:
   61  		}
   62  
   63: 		_entity.pickupMeleeWeaponAndShield(_entity.getTile());
   64  		_entity.getItems().payForAction([]);
   65  		return true;

D:\BB-Modding\vanilla\scripts\entity\tactical\skeleton.nut:
  367  		this.actor.onResurrected(_info);
  368  		this.m.IsResurrected = true;
  369: 		this.pickupMeleeWeaponAndShield(this.getTile());
  370  		this.getSkills().update();
  371  		this.m.XP /= 4;

D:\BB-Modding\vanilla\scripts\entity\tactical\enemies\zombie.nut:
  496  		this.actor.onResurrected(_info);
  497  		this.m.IsResurrected = true;
  498: 		this.pickupMeleeWeaponAndShield(this.getTile());
  499  		this.getSkills().update();
  500  		this.m.XP /= 4;

3 matches across 3 files
```

Reply:
Yeah actually that actually explains a lot. I guess just make sure if we get passed nothing we do nothing...

Reply:
No it isn't as simple as that. Vanilla passes nothing because they want the entity to spend 4 AP (the default item action cost). So, that's how I have it set up right now:
```squirrel
o.getActionCost = function( _items )
{
	if (this.m.IsMSUHandledItemAction) return 0;

	this.m.ActionSkill = null;

	local info = this.getActor().getSkills().getItemActionCost(_items);

	info.sort(@(info1, info2) info1.Skill.getItemActionOrder() <=> info2.Skill.getItemActionOrder());

	local cost = ::Const.Tactical.Settings.SwitchItemAPCost;

	foreach (entry in info)
	{
		if (entry.Cost < cost)
		{
			cost = entry.Cost;
			this.m.ActionSkill = entry.Skill;
		}
	}

	return cost;
}

o.payForAction = function ( _items )
{
	local actionCost = this.getActionCost(_items);
	this.m.Actor.setActionPoints(::Math.max(0, this.m.Actor.getActionPoints() - actionCost));
	if (_items.len() != 0) this.m.Actor.getSkills().onPayForItemAction(this.m.ActionSkill, _items);
	this.m.ActionSkill = null;
}
```
So, basically `getActionCost` returns the default cost if `_items` is an empty array. However, if it is an MSU Handled Item Action (which is a switch we ONLY use for handling AI related vanilla item actions), then it returns 0.

The main question that I am a bit confused about and would like input from you and @TaroEld here is whether we call the `onPayForItemAction` event for skills if `_items` is empty or not. Currently, I've set it to not call that, because `_items` is only empty during AI related actions and calling that event can trigger downstream changes in skills which have that event and do something based on that.